### PR TITLE
Support IAM Roles for Tasks (#346).

### PR DESCRIPTION
**Pull request by aaithal**
* Update acs client to recognize new payload types that contain IAM Role
  Credentials
* Update acs handlers to process new message types for IAM Role Credentials
* Update the api task struct to save a reference to the credentials id for
  the task
* Add the credentials manager to manage credentials
* Add the credentials endpoint to handle requests for credentials from
  containers

r? @samuelkarp @juanrhenals @richardpen @kiranmeduri

### CODE REVIEW

**Comment by samuelkarp** [comment]
:+1:

**Comment by juanrhenals** [comment]
👍  LGTM

File: agent/api/task.go [hunk 1]
```diff
@@ -462,6 +505,11 @@ func (task *Task) UpdateDesiredStatus() {
 	task.updateContainerDesiredStatus()
 }
 
+func (task *Task) SetKnownStatus(status TaskStatus) {
+	task.KnownStatus = status
```
Comment (richardpen) [inline agent/api/task.go]: Can we just use the two functions: setKnownStatus and updateKnownStatusTime which has been defined below.

File: agent/acs/handler/acs_handler.go [hunk 2]
```diff
@@ -216,13 +261,18 @@ func newDisconnectionTimer(client wsclient.ClientServer, _time ttime.Time, timeo
...
 // kinds of messages expected from ACS. It returns on server disconnection or when
 // the context is cancelled
-func startACSSession(ctx context.Context, client wsclient.ClientServer, timer ttime.Timer, args StartSessionArguments, backoff *utils.SimpleBackoff) error {
+func startACSSession(ctx context.Context, client wsclient.ClientServer, timer ttime.Timer, args StartSessionArguments, backoff *utils.SimpleBackoff, acsSessionState sessionState) error {
 	// Any message from the server resets the disconnect timeout
 	client.SetAnyRequestHandler(anyMessageHandler(timer))
 	cfg := args.Config
 
+	refreshCredsHandler := newRefreshCredentialsHandler(ctx, cfg.Cluster, args.ContainerInstanceArn, client, args.CredentialsManager, args.TaskEngine)
+	defer refreshCredsHandler.clearAcks()
```
Comment (richardpen) [inline agent/acs/handler/acs_handler.go]: should we change the order of these to:
```
client.AddRequestHandler(refreshCredsHandler.handlerFunc())
refreshCredsHandler.start()
defer refreshCredsHandler.clearAcks()
```

File: agent/api/task.go (hunk 1 above)
Comment (aaithal) [inline agent/api/task.go]: Missed this, thanks!

File: agent/acs/handler/acs_handler.go (hunk 2 above)
Comment (aaithal) [inline agent/acs/handler/acs_handler.go]: I don't think it matters. leaving it as it is.

**Comment by richardpen** [comment]
👍

**Pull request merged**